### ISSUE TITLE
How to get the the Gumbel softmax distributions over vocabulary inside GumbelSoftmaxEmbeddingHelper

### ISSUE THREAD

**Issue by XinDongol** [asyml/texar-pytorch#161]
```python
            helper_2 = tx.modules.GumbelSoftmaxEmbeddingHelper(
                # embedding=self.encoder.word_embedder,
                start_tokens=torch.tensor([101]*2).long(),
                end_token=102,
                tau=1,
                straight_through=1.)
            outputs, _, sequence_length = self.decoder(
                memory=outputs_enc[0], 
                memory_sequence_length=batch['source_length'],
                max_decoding_length=11, helper=helper_2)
```
`GumbelSoftmaxEmbeddingHelper` provide a very easy way to feed `Gumbel softmax distributions over vocabulary` to the next step. How can I get the `Gumbel softmax distributions over vocabulary` at each time step?

**Comment by gpengzhi** [comment]
Thank you for your interest in `texar-pytorch`! 

To the best of my knowledge, `outputs` here contains the outputs at all time steps. 

1. If you set `straight_through` to be `False`, the  `Gumbel softmax distribution` is fed to the next step at each time. You then can access to all the `distributions` from `outputs.sample_ids`.

2. If you want to set `straight_through` to be `True`, which means a single token with the highest probability is fed to the next step at each time. I do not think we store the `distribution` information in the case. 

Please correct me if I am wrong. @huzecong

**Comment by XinDongol** [comment]
Thanks for your quick reply. 
I checked `outputs.sample_ids` and found that if I set `straight_through=True`. The `outputs.sample_ids` is one-hot vector. 
What I want to do is to pass this one-hot vector to another model D (saying a discriminator). 

Can gradient flow through this one-hot vector from model D?

**Comment by gpengzhi** [comment]
If you set `straight_through=True`, `outputs.sample_ids` then stores the tokens with the highest probability at each time step. So the one-hot vector is expected.

In this case, `sample_ids` is computed by `(sample_ids_hard - sample_ids).detach() + sample_ids` ([here](https://github.com/asyml/texar-pytorch/blob/master/texar/torch/modules/decoders/decoder_helpers.py#L625)), where `sample_ids_hard` is the token index (one-hot vector), and `sample_ids` is the `distribution` tensor. I think the gradient will flow through the `distribution` tensor.

You might also want to wait for the feedback from @huzecong who wrote this part of the code.

**Comment by huzecong** [comment]
@gpengzhi explained it right. The definition of "straight-through" Gumbel-softmax is to use the IDs with maximum probability (hard IDs) during the forward pass, but use the softmax distribution (soft IDs) during the backward pass. Our implementation follows this behavior, so the gradient can flow through this one-hot vector.

The trick here is to use `detach`. As @gpengzhi pointed out, we compute the straight through sample IDs by `(hard_ids - soft_ids).detach() + soft_ids`, where `detach` removes the tensor from the computation graph so gradients are not computed for them. Thus, in the forward pass, the value would be `hard_ids - soft_ids + soft_ids = hard_ids`, while in the backward pass, only the non-detached `soft_ids` get used in gradient computation.

**Comment by XinDongol** [comment]
Thanks for @gpengzhi @huzecong help! I think my question is well answer and I will close this thread. 

One suggestion is that the document can be further improved if possible to give users more information. 
For example, right now, the definition of `sample_ids` is 

> The sampled results (at each step/of all steps). For example, in AttentionRNNDecoder with decoding strategy of "train_greedy", this is a torch.LongTensor of shape [batch_size, max_time] containing the sampled token indices of all steps.

I thought `sample_ids` is just the `index of tokens` when I first read this paragraph. However, it turns out that it can also be a one-hot vector which has gradient if I use gumbel helper. 

I appreciate all you guys' help and good work.  Texar is really awesome. 👍

**Comment by huzecong** [comment]
@XinDongol Thank you for pointing this out! We will update the documentation accordingly for the helpers. Thank you for supporting the Texar library!
